### DISCUSSION TITLE
x-collapse overflow hidden

### DISCUSSION THREAD

**Discussion by reefki** [alpinejs/alpine#2207] [7. Bugs (archived)]
When using `x-collapse` I found that it adds `overflow: hidden;` style to the element even when it on collapsed state. I know that it's really necessary to get a smooth transition but it makes a little mess with my dropdown inside that being chopped due to the `overflow: hidden;`.

I think it's better to use `overflow: initial;` when the element not expanded and change it back to `overflow: hidden;` when starting the transition and when the element collapsed.

I don't know if that's possible but let see what @calebporzio think.

Thanks

**Comment by SimoTod** [comment]
Overflow:hidden is required when the element is collpased otherwise it would show the content of the div on your page when the height is 0
I think you meant to suggest removing it when the element is not collapsed and after the animation right?

P.s. https://twitter.com/calebporzio/status/1430564014464159749?t=jbKbTNhbKrDA6qF-eajv2Q&s=19

**Reply by reefki** [reply]
@SimoTod Yeah that's exactly what I meant

Also, thanks for the heads up on Caleb's tweet. My apologies.

**Reply by SimoTod** [reply]
It makes sense, I think you can send a Pr for this, it will probably get approved.

**Comment by areski** [comment]
Anyone found a temporary solution to this issue?

**Reply by areski** [reply]
In the meantime, I found this solution which works nicely for me:
`<div x-show="show_search_filter"
        x-collapse
        x-on:transitionend="show_search_filter ? $el.style.overflow='visible' : ''"
        >`

**Reply by AnthonyANI** [reply]
For those still struggling with this and x-on:transitionend isn't always working, try x-on:focusin instead as a workaround.
```html
<div x-cloak x-show="expanded" x-collapse x-on:focusin="expanded ? $el.style.overflow = null : ''">
```

**Reply by ekwoka** [reply]
Can you show more context to what is actually happening?

I can't image what situation would need to have it be....hidden but also visible?

**Reply by AnthonyANI** [reply]
Hey @ekwoka, glad to help. I actually have somewhat of the opposite issue that the OP does, though there are similar issues that have the same issue I do [2208](https://github.com/alpinejs/alpine/issues/2208), [3549](https://github.com/alpinejs/alpine/issues/3549).

The issue I'm having is when the `overflow: hidden` style is not removed after the element finishes showing/opening. I recorded the problem on the Alpine.js demo page: 

https://github.com/alpinejs/alpine/assets/1095768/97479f20-4651-4105-ab03-842d20dfa655

This causes issues with custom dropdrops like Select2, Selectize, Choices.js that use an absolute-positioned div to simulate a dropdown window. Those are cut off because they're children of the x-collapse element which has `overflow: hidden`.

I can see the OP's situation as well however if using `x-collapse.min` and trying to show a dropdown in that area. The only way I can think to solve that is to have x-collapse on a separate child div. (One div, two child divs: top div for the header with the dropdown and bottom with the `x-collapse`).

Thanks!

**Reply by ekwoka** [reply]
Okay, that makes more sense than the issue of it not removed when it's collapsed.

**Reply by AnthonyANI** [reply]
Just in case it's related, I added an `x-on:transitionend="console.log('done transitioning')"` to re-test it since I was having issues. And it also does not always fire. Not sure if that would be related to this issue or not. ~~The problems appear to occur independently of each other, though both intermittent. Sometimes the `overflow: hidden` will be removed but the console log never happened or vice versa.~~

I did test that with Tailwind, which may have made it worse. I created a CodePen to try it without Tailwind and I couldn't get the problems to occur independently of each other. When the `overflow: hidden` failed to be removed after expanding, the `x-on:transitionend` also did not fire. https://codepen.io/AnthonyANI/full/dyQpRYN

**Reply by ekwoka** [reply]
Yes, I found transitionend to also be questionable on collapse.

**Reply by willvincent** [reply]
Just wanted to weigh in that this is still very much an issue even after [PR #3243](https://github.com/alpinejs/alpine/pull/3243). What's really frustrating is that it's sort of intermittent, and not a zoom issue as that PR suggested.

I've omit `x-collapse` for the time being, which takes some of the gloss off the UI, but at least my ChoicesJS select field continues to function as expected within the collapsible div. Guess I may try that focusin workaround above, but it sure seems like this should just get fixed properly.

**Reply by code23-isaac** [reply]
I'm using x-collapse on a container with a max-height and overflow-y scrolling:

```      
            <div x-show="isOpen" x-collapse class="overflow-y-auto lg:max-h-60">
                <div class="pb-5">
                    @foreach ($categoryList as $category)
                    @include('partials.shop-filter-categories', ['category' => $category])
                    @endforeach
                </div>
            </div>
```

By console logging the height comparison surrounding the overflow style reset (index.js:60):

```
if (el.getBoundingClientRect().height == full) {
   el.style.overflow = null
}
```

`console.log(el.getBoundingClientRect().height, full);`

it seems the heights don't match every time in chrome, and in safari they never match:

![Screenshot 2023-12-15 at 10 57 08](https://github.com/alpinejs/alpine/assets/38762024/72ddb98f-d56d-4118-b855-7339bd49b225)
![Screenshot 2023-12-15 at 11 02 14](https://github.com/alpinejs/alpine/assets/38762024/1e687979-6285-4f0f-8d3f-784c32a7cddb)

So for this scenario the plugin is broken in Safari as the overflow-hidden style is never removed, and intermittently broken in Chrome.

Is the height comparison around the nullifying of the overflow style required? Removing it fixes the issue, but maybe causes visual bugs on slower transitions?

**Comment by grey4owl** [comment]
I created a PR(https://github.com/alpinejs/alpine/pull/3949) that solves this.. I also added a optional modifier `.overflow` which preserve overflow after expanding if you want, and I changed overflow value from `hidden` to `clip` because it supports shadows and offset with `overflow-clip-margin` property..

**Comment by jackmurraybrown** [comment]
I managed to solve this by applying `overflow-y: auto !important` to the element with `x-collapse`.
